Implement a -cmi-file option

shindere opened the pull request:
This is an attempt at fixing issue #9717.

The provided fix consists in introducing a `-cmi-file` command-line
option that lets one specify against which binary interface file an
implementation should be compiled.

If the option is provided, the lookup for the source (.mli) interface
besides to the .ml module is skipped.

If the option is not specified the compiler behaves as before.

To follow-up on the example given in #9717, it is now possible to
compile `impl/a.ml` against `build/a.cmi` as follows:

```
ocamlc -c impl/a.ml -cmi-file build/a.cmi -o build/a.cmo
```

From a functional point of view, is such an implementation good enough for
clever build systems that can easily specify complete filenames rather than
just directories?

Would it also be good to be able to specify just the directory
where to look for .cmi files? IF so, dos that require a dedicated option,
or shall the normal include path be used? In that last case,
should there be a dedicated option to tell teh compiler to look
in the include dire for .cmi files, or should the default behaviour
simply be extended?

Assuming the feature is considered satisfactory, I must say I am not
very happy about its implementation because the pattern-matching on
`!Clflags.cmi_file`is done twice and because `sourceintf` is computed
even when not required. I wrote things that way because that looked the
less invasive way.

shindere commented:
As a follow-up to an off-line discussion and review by @damiendoligez, I
just rebased and slightly simplified the code.

The manual has also been updated and a Change entry has been added.

shindere commented:
@Octachron you may also want to have a look, especially to the documentation.

dbuenzli commented on manual/src/cmds/unified-options.etex:
```diff
@@ -116,6 +116,11 @@ Pass the given option to the C compiler and linker.
 
+\notop{%
+\item["-cmi-file" \var{filename}]
+Use the given interface file to type-check the ML source file to compile.
```
It would be nice to indicate what's the behaviour if the option is unspecified, i.e. IIRC that the `.cmi` will be looked up in the load path but that it is expected that there is an existing `.mli` file at the same place as the .ml source.

damiendoligez replied:
more like: **if** there is an `.mli` a the same place as the `.ml`, **then** the `.cmi` will be looked up in the load path, **else** the `.cmi` will be output by the compiler (next to the `.cmo` or `.cmx`).

dbuenzli reviewed (approved):
> From a functional point of view, is such an implementation good enough for clever build systems that can easily specify complete filenames rather than just directories?

It's not good enough, it's perfect.

> Would it also be good to be able to specify just the directory where to look for .cmi files?

Personally I don't need it. If no-one has a use for it, I think it's better not to clutter the cli with more options.

Thanks for the work on the feature.

dbuenzli commented on typing/typemod.ml:
```diff
@@ -3007,13 +3007,17 @@ let type_implementation sourcefile outputprefix modulename initial_env ast =
         let sourceintf =
           Filename.remove_extension sourcefile ^ !Config.interface_suffix in
-        if Sys.file_exists sourceintf then begin
+        if Sys.file_exists sourceintf || (!Clflags.cmi_file <> None) then begin
```
If you swap the arguments of the `||` you can avoid one syscall here.

shindere commented:
Daniel Bünzli (2022/02/07 12:17 -0800):
> @dbuenzli approved this pull request.

Thanks!

> > From a functional point of view, is such an implementation good enough for clever build systems that can easily specify complete filenames rather than just directories?
>
> It's not good enough, it's perfect.

Thanks, once more.

> > Would it also be good to be able to specify just the directory where to look for .cmi files?
>
> Personally I don't need it. If no-one has a use for it, I think it's
> better not to clutter the cli with more options.

Yeah, I gave it a thought and it feels that even for build-systems using
make, like the one of the compiler, the currently implemented option is
enough.

I also intend to submit a separate PR that adds this option to ocamldep /
ocamlc -depend, and also -cmx-file and -cmo-file options so that it
becomes possible to customize all the aspects of the generated
dependency rules.

> Thanks for the work on the feature.

My pleasure. It will be helpful in many contexts, I think.

> > @@ -116,6 +116,11 @@ Pass the given option to the C compiler and linker.
>  directory \var{dir}. \comp{(See the "-custom" option.)}
>  }%notop
>
> +\notop{%
> +\item["-cmi-file" \var{filename}]
> +Use the given interface file to type-check the ML source file to compile.
>
> It would be nice to indicate what's the behaviour if the option is
> unspecified, i.e. IIRC that the `.cmi` will be looked up in the load
> path but that it is expected that there is an existing `.mli` file at
> the same place as the .ml source.

I could do that, yes. I think by default the `.cmi` and `.mli` files are
only looked for in the directory where the `.ml` file resides, rather
than in the complete load path.

I also need to update the manpages, which seem to need a verification
because other options seem to be missing ther.

dbuenzli commented:
> I think by default the `.cmi` and `.mli` files are only looked for in the directory where the `.ml` file resides, rather than in the complete load path.

No, the `.cmi` is looked in the load path and the `.mli` in the directory where the `.ml` file resides (you can see it in the diff of the PR).

shindere commented:
Daniel Bünzli (2022/02/07 12:20 -0800):
> @dbuenzli commented on this pull request.
>
>
>
> > @@ -3007,13 +3007,17 @@ let type_implementation sourcefile outputprefix modulename initial_env ast =
>        end else begin
>          let sourceintf =
>            Filename.remove_extension sourcefile ^ !Config.interface_suffix in
> -        if Sys.file_exists sourceintf then begin
> +        if Sys.file_exists sourceintf || (!Clflags.cmi_file <> None) then begin
>
> If you swap the arguments of the `||` you can avoid one syscall here.

that's a good idea! Thanks!

I just pushed the update and credited you in the Changes entry. :)

shindere commented:
Daniel Bünzli (2022/02/07 12:29 -0800):
> > I think by default the `.cmi` and `.mli` files are only looked for in the directory where the `.ml` file resides, rather than in the complete load path.
>
> No, the `.cmi` is looked in the load path and the `.mli` in the
> directory where the `.ml` file resides (you can see it in the diff of
> the PR).

Yes, you are right of course. Sorry.

damiendoligez reviewed (approved):
The code looks good and the feature seems to satisfy everyone.

Octachron commented:
Before I forget:  `man/ocamlc.1` and `man/ocamlopt.1` need to be updated.

shindere commented:
Damien Doligez (2022/02/09 07:18 -0800):
> @damiendoligez approved this pull request.
> 
> The code looks good and the feature seems to satisfy everyone.
> 
> 
> > @@ -116,6 +116,11 @@ Pass the given option to the C compiler and linker.
>  directory \var{dir}. \comp{(See the "-custom" option.)}
>  }%notop
>  
> +\notop{%
> +\item["-cmi-file" \var{filename}]
> +Use the given interface file to type-check the ML source file to compile.
> 
> more like: **if** there is an `.mli` a the same place as the `.ml`,
> **then** the `.cmi` will be looked up in the load path, **else** the
> `.cmi` will be output by the compiler (next to the `.cmo` or `.cmx`).

What do you mean, here? Your suggestion seems to describe the behaviour
when the option is not given. Do you menat to include this in the
documentation? At this place?

dbuenzli commented:
> Do you mean to include this in the documentation? At this place?

Yes and yes. FWIW that's what I also meant.

shindere commented:
Florian Angeletti (2022/02/09 07:21 -0800):
> Before I forget:  `man/ocamlc.1` and `man/ocamlopt.1` need to be
> updated.

Done

shindere committed:
Implement a -cmi-file option for ocamlc and ocamlopt

This lets the user specify against which interface a source module
should be type-checked.

When this option is specified, the usual check for a corresponding
.mli file is skipped.

shindere commented:
Daniel Bünzli (2022/02/11 08:59 +0000):
> > Do you mean to include this in the documentation? At this place?
> 
> Yes and yes. FWIW that's what I also meant.

Done so.

Here is what I just added to the manual:

```
When this option is not specified, the compiler looks for a \var{.mli} file
with the same base name than the implementation it is compiling and in the
same directory. If such a file is found, the compiler looks for a
correspoinding \var{.cmi} file in the included directories and reports an
error if it fails to find one.
```

shindere commented:
Just pushed e600956647a8647c494d74d9428f5903e6b7245c to fix the Changes entry
